### PR TITLE
[TEP-0153] Migration of Tekton Results to V1 APIs of Pipelines

### PR DESCRIPTION
**Pull request by khrm**
TEP discussing the approach to take to introduce the support for v1 APIs of pipelines.

### CODE REVIEW

**Comment by khrm** [comment]
@tektoncd/results-maintainers @tektoncd/results-collaborators

**Comment by adambkaplan** [comment]
/assign

@dibyom sure, I can be put down as a reviewer.

**Comment by adambkaplan** [comment]
The current draft captures some of the rhetorical questions I posed in https://github.com/tektoncd/results/issues/303#issuecomment-1655730555.

IMO we should first decide if we make the reconcile of v1 objects an opt-in/opt-out feature, or a hard-coded requirement. The latter is simpler, but imposes a stronger restriction on the versions of Tekton Pipelines that Results can support From a community perspective, I don't see this being an issue as our oldest LTS release of Pipelines includes v1 TaskRun and PipelineRun APIs (see [here](https://tekton.dev/vault/pipelines-v0.41.x-lts/pipeline-api/#tekton.dev%2fv1)).

If we make the v1 reconcile an opt-in feature, we'll need to add some version skew/glue code in the watcher. This won't be very daunting because the core reconcile logic of watcher codebase uses dynamic clients and has a lot of reusable functions. Over the long run, though, it will be tech debt that will be carried until the v1beta1 API is disabled for good.

I'd like other @tektoncd/results-collaborators and @tektoncd/results-maintainers to weigh in as well before we proceed.

File: teps/0141-v1-migration-for-tekton-results.md
Comment (chitrangpatel) [inline teps/0141-v1-migration-for-tekton-results.md:76]: In Chains, we rely on the conversion functions to convert between `v1beta1` and `v1`. Could you not do something similar?

Our plan is to watch `v1` by default. We will provide the cluster operators to configure chains to generate `v1` or `v1beta` based provenance. Internally, we will migrate all our functions to run with `v1`. There are some deprecated fields like piplineResources that strictly need `v1beta1` so our plan is to just convert the tekton object to v1beta1 for extracting the necessary information out.

Comment (khrm) [reply teps/0141-v1-migration-for-tekton-results.md:76]: Will you convert stored v1beta1 to v1 also? Do you store yaml in provenance? Will you convert the stored one?

Comment (khrm) [reply teps/0141-v1-migration-for-tekton-results.md:76]: At present, we just serve base64 encoded blob in a JSON field to the client without checking whether the stored object is v1 or v1beta1.

Comment (adambkaplan) [reply teps/0141-v1-migration-for-tekton-results.md:76]: My recommendation:

1. Results updates to reconcile `v1` object types.
2. Results API server is updated to check the type that is being served by the `Record`. If it is a v1beta1 object type, we must decode the object, convert to v1, then re-encode before sending over the wire.
3. We enhance the `tkn-results ` cli plugin with a `convert` command, that converts a `v1beta1` api object to `v1`. This could be packaged ina container image and executed as a `Job` on a cluster.

File: teps/0141-v1-migration-for-tekton-results.md
Comment (adambkaplan) [inline teps/0141-v1-migration-for-tekton-results.md:83]: nit - clarify language:

```suggestion
* What happens when v1beta1 conversion functions are removed?
```

**Review by khrm — COMMENTED** [review]
/assign @enarha @sayan-biswas

**Review by khrm — COMMENTED** [review]
@sayan-biswas @enarha @avinal 

I have added all the points and concerns that we discussed in the WG call. I think we can try to merge this now.

**Comment by khrm** [comment]
I need to rebase again I think

**Review by avinal — APPROVED** [review]
Looks good to me.

**Comment by adambkaplan** [comment]
@khrm please rebase.

This _**needs**_ to happen, so as a `proposed` TEP this should merge. But to get to `implementable`, there are a lot of decisions that need to be actively discussed in the working group. I would be okay with having more than one follow up PR on this TEP that brings forward decision points that are made in the Results WG.

**Comment by pritidesai** [comment]
API WG - @dibyom to review this! Thanks!

**Review by sayan-biswas — APPROVED** [review]

**Comment by avinal** [comment]
Is reviewed and ready to be merged? cc. @pritidesai @dibyom

**Comment by khrm** [comment]
@avinal I have rebased this.

**Comment by avinal** [comment]
Everything seems fine. We have required approvals as well. 
/lgtm

**Pull request merged**
